File: 1020. Number of Enclaves/solution.py
from typing import List

class Solution:
	def numEnclaves(self, grid: List[List[int]]) -> int:
		def propogate(x, y):
			if ((x < 0 or x > row_len - 1) or
				(y < 0 or y > col_len - 1) or
				grid[x][y] == 0):
				return
			grid[x][y] = 0
			propogate(x - 1, y)
			propogate(x + 1, y)
			propogate(x, y + 1)
			propogate(x, y - 1)

		row_len = len(grid)
		col_len = len(grid[0])

		for i in range(row_len):
			for j in range(col_len):
				if ((i == 0 or j == 0 or i == (row_len - 1) or j == (col_len - 1)) and grid[i][j] == 1):
					propogate(i, j)

		return sum(sum(row) for row in grid)

File: 1020. Number of Enclaves/solution.c
/**
 * @brief	Solution uses a Depth First Search approach to eliminate the island clusters
 * 			that touched the edge of the map. The remaining land boxes are then summed up
 * 			and returned.
*/

void    dfs_sink(int **grid, int x, int y, int max_row, int max_col){
	if (x == -1 || y == -1 || x == max_row || y == max_col || grid[x][y] == 0)
        return ;
    grid[x][y] = 0;
    dfs_sink(grid, x + 1, y, max_row, max_col);
    dfs_sink(grid, x, y + 1, max_row, max_col);
    dfs_sink(grid, x - 1, y, max_row, max_col);
    dfs_sink(grid, x, y - 1, max_row, max_col);
}

void    sink_borders(int **grid, int row, int col){
    for (int i = 0; i < col; i++){
        dfs_sink(grid, 0, i, row, col);
        dfs_sink(grid, row - 1, i, row, col);
    }
    for (int j = 0; j < row; j++){
        dfs_sink(grid, j, 0, row, col);
        dfs_sink(grid, j, col - 1, row, col);
    }
}

// Driver function
int numEnclaves(int** grid, int gridSize, int* gridColSize){
    int row_len = gridSize, col_len = gridColSize[0];
    int i = -1, j = -1, sum = 0;

    sink_borders(grid, row_len, col_len);
    while (++i < row_len){
        j = -1;
        while (++j < col_len){
            sum += grid[i][j];
        }
    }
    return (sum);
}